src/main/kotlin/com/rstudio/shinycannon/Main.kt

```diff
@@ -219,39 +219,38 @@ class ShinySession(val sessionId: Int,
     }
 }
 
-fun nowMs() = Instant.now().toEpochMilli()
+fun nowMs() = Instant.now().millis
 
 fun PrintWriter.printCsv(vararg columns: Any) {
     this.println(columns.joinToString(","))
     this.flush()
 }
 
 class Stats() {
-    enum class State { RUN, DONE, FAIL }
     enum class Transition { RUNNING, FAILED, DONE }
 
-    val stats = ConcurrentHashMap(mapOf(
-            State.RUN to 0,
-            State.DONE to 0,
-            State.FAIL to 0
-    ))
+    var run = 0
+    var done = 0
+    var fail = 0
 
     fun transition(t: Transition) {
-        stats.replaceAll { k, v ->
-            when (Pair(t, k)) {
-                Pair(Transition.RUNNING, State.RUN) -> v + 1
-                Pair(Transition.DONE, State.RUN) -> v - 1
-                Pair(Transition.DONE, State.DONE) -> v + 1
-                Pair(Transition.FAILED, State.RUN) -> v - 1
-                Pair(Transition.FAILED, State.FAIL) -> v + 1
-                else -> v
+        synchronized(this) {
+            when (t) {
+                Transition.RUNNING -> run++
+                Transition.DONE -> {
+                    done++
+                    run--
+                }
+                Transition.FAILED -> {
+                    fail++
+                    run--
+                }
             }
         }
     }
 
     override fun toString(): String {
-        val copy = stats.toMap()
-        return "Running: ${copy[State.RUN]}, Failed: ${copy[State.FAIL]}, Done: ${copy[State.DONE]}"
+        return "Running: ${run}, Failed: ${fail}, Done: ${done}"
```

Comment:
Please wrap in `synchronized(this)` as well

Reply:
Thanks, I ended up making a `@Synchronized` accessor that returns a consistent view of the counts so that synchronization for reads only happens in one place.